### DISCUSSION TITLE
SuperRefine being executed on invalid data

### DISCUSSION THREAD

**Discussion by ali-idrizi** [colinhacks/zod#1253] [General]
Using version `3.17.3`, considering:

```
const validator = z
  .object({
    name: z.string().min(1, 'Name is required'),
  })
  .superRefine((data, ctx) => {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'superRefine error',
    })
  })
```

I am getting the error "superRefine error" when `name` is `''`, but not when `name` is `undefined`.

`validator.parse({name: ''})` returns:

```
[
  {
    code: 'too_small',
    minimum: 1,
    type: 'string',
    inclusive: true,
    message: 'Name is required',
    path: ['name'],
  },
  {
    code: 'custom',
    message: 'superRefine error',
    fatal: true,
    path: [],
  },
]
```

However when the name is not provided at all, so `validator.parse({})`, I get:

```
[
  {
    code: 'invalid_type',
    expected: 'string',
    received: 'undefined',
    path: ['name'],
    message: 'Required',
  },
]
```
As you can see the first example run the `superRefine`, while the second one did not. Is this intended? It doesn't seems consistent. I read the readme and there is no mention on what condition `superRefine` executes. I would like `superRefine` not to execute at all when the data isn't parsed successfully, is there an easy way to accomplish this?

**Comment by nullndr** [comment]
The `.min()` method is actually a _refinement type_. Poorly speak, it is just a ZodString with a `.superRefine()` method.

The `.superRefine()` method does not throw when the parsing fails, instead it continue in order to check for others refinements: see [here](https://github.com/colinhacks/zod#abort-early).

The first example parse successfully the `""` literal because it is truly a string, fails the `.min()` and will continue to your `.superRefine()`.

This does not happens on your second example, simply because `validator` fails to parse `undefined` as a string. The parse fails, and the `.min()` refinement is not executed, so it is not chained to your second `.superRefinement()`.

**Reply by nullndr** [reply]
@ali-idrizi does this answer your question?
